src/umessage.rs

```diff
@@ -17,3 +17,151 @@ mod umessagetype;
 pub use umessagebuilder::*;
 
 pub use crate::up_core_api::umessage::UMessage;
+
+use crate::{UAttributesError, UPayloadFormat};
+use protobuf::{well_known_types::any::Any, Message, MessageFull};
+
+#[derive(Debug)]
+pub enum UMessageError {
+    AttributesValidationError(UAttributesError),
+    DataSerializationError(protobuf::Error),
+    PayloadError(String),
+}
+
+impl std::fmt::Display for UMessageError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::AttributesValidationError(e) => f.write_fmt(format_args!(
+                "Builder state is not consistent with message type: {}",
+                e
+            )),
+            Self::DataSerializationError(e) => {
+                f.write_fmt(format_args!("Failed to serialize payload: {}", e))
+            }
+            Self::PayloadError(e) => f.write_fmt(format_args!("UMessage payload error: {}", e)),
+        }
+    }
+}
+
+impl std::error::Error for UMessageError {}
+
+impl From<UAttributesError> for UMessageError {
+    fn from(value: UAttributesError) -> Self {
+        Self::AttributesValidationError(value)
+    }
+}
+
+impl From<protobuf::Error> for UMessageError {
+    fn from(value: protobuf::Error) -> Self {
+        Self::DataSerializationError(value)
+    }
+}
+
+impl From<&str> for UMessageError {
+    fn from(value: &str) -> Self {
+        Self::PayloadError(value.into())
+    }
+}
+
+impl UMessage {
+    /// Insert a protobuf object into the payload field of a `UMessage` object.
+    ///
+    /// This function is used to encapsulate a strongly-typed data object into a `UMessage`,
+    /// which allows for more generic data handling. Target payload encoding can be controlled
+    /// to be one of `UMessage::UPayloadFormat`, and will only work for `UPayloadFormat::UPAYLOAD_FORMAT_PROTOBUF`
+    /// or `UPayloadFormat::UPAYLOAD_FORMAT_PROTOBUF_WRAPPED_IN_ANY`.
+    ///
+    /// # Type Parameters
+    ///
+    /// * `T`: The type of the data to be packed.   
+    ///
+    /// # Parameters
+    ///
+    /// * `data`: The data to pack.
+    /// * `payload_format`: Optional `UMessage::UPayloadFormat`, defining serialization to use. Will default to UPayloadFormat::UPAYLOAD_FORMAT_PROTOBUF.
+    ///
+    /// # Returns
+    ///
+    /// * Ok(()): if the operation succeeded
+    ///
+    /// # Errors
+    ///
+    /// * Err(`UMessageError`) if the serialization process fails, for example if the payload format is unsupported.
+    pub fn insert_protobuf_payload<T: MessageFull>(
```

Comment:
FMPOV we do not need this function and should instead keep the builder's corresponding functionality. Or am I missing something?

Reply:
The thought was that it would be useful to modify the payload of an existing UMessage object - so these methods went into UMessage, not least because they are mirroring the extract_payload() methods that imo definitely belong in UMessage.

If that is not a valid use case at all, then I agree and the logic can move into UMessageBuilder.

Reply:
IMHO even though we're not forbidding anyone from modifying a `UMessage`, since we're publicly exposing the protobuf type, I like the idea of making the "obvious" way to make a `UMessage` very much a one-way-of-doing-things process.

So my stance would be that we should not go out of our way to support inserting a protobuf payload after the fact, folks should make use of the `UMessageBuilder` to just make the message they need with the payload they need.

> not least because they are mirroring the extract_payload() methods that imo definitely belong in UMessage.

Sorry, I'm not familiar with what you're writing about here. I agree there should be some "nice" way of getting at the payload.

Reply:
Ok, bowing to majority opinion - moved the code to UMessageBuilder.